### PR TITLE
Implement the Folder repository

### PR DESCRIPTION
**Pull request by MathieuxC**
This PR add an implementation of the folder repository to communicate with firebase.

**Structure in the Database**
The folders are stored in a collection `folders` of the database. This collection contains documents representing the actual folders. They have a field `ownerId` that allows retrieving easily all the folders associated to a user.
Additionaly to the fields with the attribute of folders, each document contains a `files` collection, which itself contains the files.

Capture d'écran 2024-10-14 194018
es.

**Created methods**
All methods from the repository interface have been created. In paritcular, we can now add a new folder to the database, update or remove an existing one, retrieve all the folders associated to a user, and create new folders/files with a unique ID.
Concerning this last part, the repository has been updated to have separate methods `getNewFolderUid` and `getNewFileUid`, instead of a unique `getNewUid`.

**Other changes**
Apart from creating the repository, this PR also introduce to small changes.

1. The creation of a factory for the view model.
2. The addition of a field `id` to the MyFile data class. It represent the unique id of the file in the database (~the name of the document). All usage have been updated accordingly.

### CODE REVIEW

**Comment by abj10** [comment]
LGTM

**Comment by MathieuxC** [comment]
While working using this implementation of the repository, I encountered several difficulties which led me into making important change to the class. Those do not affect the API of the repo, except for the suppression of method `getNewFileUid`, and renaming `getNewFolderUid` into `getNewUid`.

However, there are important changes on the way data is stored on the database. Creating an intern collection of files in each folder (both file and folder are documents on firebase here), I stored the files into a list directly in the document. The files themselves are represented by a `Map<String, String>`.

Capture d'écran 2024-10-16 221952

File: app/src/androidTest/java/com/github/se/eduverse/ui/folder/DisplayFolderTest.kt
Comment (Fayroth) [inline app/src/androidTest/java/com/github/se/eduverse/ui/folder/DisplayFolderTest.kt:1]: LGTM

File: app/src/test/java/com/github/se/eduverse/repository/FolderRepositoryTest.kt
Comment (Fayroth) [inline app/src/test/java/com/github/se/eduverse/repository/FolderRepositoryTest.kt:1]: LGTM

File: app/src/test/java/com/github/se/eduverse/viewmodel/FolderViewModelTest.kt
Comment (Fayroth) [inline app/src/test/java/com/github/se/eduverse/viewmodel/FolderViewModelTest.kt:1]: LGTM

Comment (Fayroth) [inline app/src/test/java/com/github/se/eduverse/viewmodel/FolderViewModelTest.kt:1]: LGTM

File: app/src/test/java/com/github/se/eduverse/repository/FolderRepositoryTest.kt
Comment (Fayroth) [inline app/src/test/java/com/github/se/eduverse/repository/FolderRepositoryTest.kt:1]: LGTM

File: app/src/main/java/com/github/se/eduverse/viewmodel/FolderViewModel.kt
Comment (Fayroth) [inline app/src/main/java/com/github/se/eduverse/viewmodel/FolderViewModel.kt:1]: The ViewModel properly initializes by trying to load user folders, handling exceptions gracefully, good use of MutableStateFlow to handle the active folder and list of folders, which provides reactive data management, implementation of Factory is well-structured for easy instantiation.

Functions like addFolder, deleteFolder, updateFolder, and renameFolder are well-defined and handle the expected flow of operations.
Methods like selectFolder, addFile, and deleteFile are straightforward and effective.

Good use of logging for error handling.

Consider using sealed classes for result handling (Success/Error) to reduce reliance on logging for flow control.

File: app/src/main/java/com/github/se/eduverse/ui/folder/DisplayFolder.kt
Comment (Fayroth) [inline app/src/main/java/com/github/se/eduverse/ui/folder/DisplayFolder.kt:1]: The screen provides a well-organized layout. UI logic like sorting can be easily triggered with the help of mutable states.

File: app/src/main/java/com/github/se/eduverse/repository/FolderRepositoryImpl.kt
Comment (Fayroth) [inline app/src/main/java/com/github/se/eduverse/repository/FolderRepositoryImpl.kt:1]: The methods for interacting with Firestore (CRUD) are well-implemented, with callbacks to handle success and failure cases.
Mapping functions (fileToMap, mapToFile) are clean and modular, facilitating easy conversion between Firestore data and app models.
Proper use of exception handling with clear, descriptive logging messages.
Implementation of filter conversion functions (stringToFilter, filterToString) ensures correct mapping between enum and database values.
Ensure that Firestore rules enforce correct user access to the folders collection, given user-specific data retrieval.

File: app/src/main/java/com/github/se/eduverse/repository/FolderRepository.kt
Comment (Fayroth) [inline app/src/main/java/com/github/se/eduverse/repository/FolderRepository.kt:1]: LGTM

File: app/src/main/java/com/github/se/eduverse/model/folder/MyFile.kt
Comment (Fayroth) [inline app/src/main/java/com/github/se/eduverse/model/folder/MyFile.kt:1]: LGTM

File: app/src/androidTest/java/com/github/se/eduverse/ui/folder/DisplayFolderTest.kt
Comment (Fayroth) [inline app/src/androidTest/java/com/github/se/eduverse/ui/folder/DisplayFolderTest.kt:1]: LGTM

**Review by Fayroth — APPROVED** [review]
The project demonstrates a well-structured implementation, with a clear separation of concerns across different components. The design patterns used help maintain clean and organized code, which makes it easier to understand and extend. The overall architecture seems to follow standard best practices, and there is an effective use of dependency injection, ensuring that components are decoupled and testable.

The project structure is logical, with appropriately named packages and classes that indicate their responsibilities. The implementation of the ViewModel pattern is particularly effective, as it helps manage UI-related data efficiently and promotes a responsive user interface. The use of repositories to handle data access is also a solid design choice, ensuring that data-related logic is isolated from the rest of the codebase.

Overall, the project is well-organized and appears to have been developed with maintainability in mind. The clean design and thoughtful separation of concerns suggest a scalable and robust implementation.

**Pull request merged**